### PR TITLE
Fix GEOmetrics Image Recon example

### PR DESCRIPTION
**Pull request by Jean-Francois-Lafleche**
### Description
Get the GEOmetrics example working

### Related Issues
#61 

### Tasks
- [x] Get main train.py script working with decreasing loss
- [x] Add resume feature to training script
- [x] Get train_auto_encoder.py and train.py with latent-loss training
- [x] Get eval.py working
- [x] Update README
- [x] Achieve good results after completing training

### Tests
OS: Ubuntu 18.04
CUDA: 10.2
Python 3.6
GPU: RTX 2080Ti
Driver: 440.44

##### Training
- `python train_auto_encoder.py --shapenet-root <path/to/ShapeNet> \
    --categories <category list> \
    --cache-dir cache/` trained for full 50 epochs.
- `python train.py --shapenet-root <path/to/ShapeNet> \
    --shapenet-images-root <path/to/ShapeNetImages> \
    --categories <category list> \
    --cache-dir cache/` trained for 300 epochs (no latent loss).
- `python train.py --shapenet-root <path/to/ShapeNet> \
    --shapenet-images-root <path/to/ShapeNetImages> \
    --categories <category list> \
    --cache-dir cache/ --latent_loss` trained for 600 epochs.
- `python train.py --shapenet-root <path/to/ShapeNet> \
    --shapenet-images-root <path/to/ShapeNetImages> \
    --categories <category list> \
    --cache-dir cache/ --latent_loss --resume best` tested resume best.
- `python train.py --shapenet-root <path/to/ShapeNet> \
    --shapenet-images-root <path/to/ShapeNetImages> \
    --categories <category list> \
    --cache-dir cache/ --latent_loss --resume recent` tested resume recent.

##### Evaluating
- `python eval_auto_encoder.py --shapenet-root <path/to/ShapeNet> \
    --categories <category list> \
    --cache-dir cache/`
image

- `python eval.py --shapenet-root <path/to/ShapeNet> \
    --shapenet-images-root <path/to/ShapeNetImages> \
    --categories <category list> \
    --cache-dir cache/`
image

### CODE REVIEW

File: examples/ImageRecon/GEOmetrics/README.md [hunk 1]
```diff
@@ -1,31 +1,38 @@
...
 ### Training the network:
 
-To train the system call
-```
-python train.py
+#### Option 1 : No latent embedding loss (faster to train)
+```bash
+python train.py --shapenet-root <path/to/ShapeNet> \
+    --shapenet-rendering-root <path/to/ShapeNetRendering> \
+    --cache-dir cache/
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/README.md:22]: Can we add detail on the purpose of `--cache-dir` and the duration of training ? (maybe with the step of the preprocessing)

File: examples/ImageRecon/GEOmetrics/architectures.py [hunk 2]
```diff
@@ -244,32 +233,27 @@ def reset_parameters(self):
...
+
+        side_2 = support[:, side_len:]
+        output = torch.cat((side_1, side_2), dim=1)
+
         if self.bias is not None:
             output = output + self.bias
         return output
 
     def __repr__(self):
         return self.__class__.__name__ + ' (' \
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/architectures.py]: Not very important but this can fit into one line

File: examples/ImageRecon/GEOmetrics/eval.py [hunk 3]
```diff
@@ -12,147 +12,142 @@
...
+
 parser = argparse.ArgumentParser()
-parser.add_argument('-expid', type=str, default='Direct', help='Unique experiment identifier.')
-parser.add_argument('-device', type=str, default='cuda', help='Device to use')
-parser.add_argument('-categories', type=str,nargs='+', default=['chair'], help='list of object classes to use')
-parser.add_argument('-vis', action='store_true', help='Visualize each model while evaluating')
-parser.add_argument('-f_score', action='store_true', help='compute F-score')
-parser.add_argument('-batchsize', type=int, default=1, help='Batch size.')
+parser.add_argument('--shapenet-root', type=str, help='Root directory of the ShapeNet dataset.')
+parser.add_argument('--shapenet-rendering-root', type=str, help='Root directory of the ShapeNet Rendering dataset.')
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/eval.py]: I called it `shapenet-images-root` in PR #72 we should be consistent between examples, what do you prefer ?

File: examples/ImageRecon/GEOmetrics/eval.py [hunk 4]
```diff
@@ -12,147 +12,142 @@
...
+parser.add_argument('--cache-dir', type=str, default=None, help='Path to write intermediate representation to.')
+parser.add_argument('--expid', type=str, default='Direct', help='Unique experiment identifier.')
+parser.add_argument('--device', type=str, default='cuda', help='Device to use')
+parser.add_argument('--categories', type=str, nargs='+', default=['chair'], help='list of object classes to use')
+parser.add_argument('--no-vis', action='store_true', help='Turn off visualization of each model while evaluating')
+parser.add_argument('--f_score', action='store_true', help='compute F-score')
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/eval.py]: please use `--f-score`

File: examples/ImageRecon/GEOmetrics/train.py [hunk 5]
```diff
@@ -14,344 +14,338 @@
...
-parser.add_argument('-lr', type=float, default=1e-4, help='Learning rate.')
-parser.add_argument('-val-every', type=int, default=5, help='Validation frequency (epochs).')
-parser.add_argument('-batch_size', type=int, default=5, help='batch size')
-parser.add_argument('-print-every', type=int, default=20, help='Print frequency (batches).')
-parser.add_argument('-latent_loss', action='store_true', help='indicates latent loss should be used')
-parser.add_argument('-logdir', type=str, default='log', help='Directory to log data to.')
-parser.add_argument('-save-model', action='store_true', help='Saves the model and a snapshot \
-	of the optimizer state.')
+parser.add_argument('--shapenet-root', type=str, help='Root directory of the ShapeNet dataset.')
+parser.add_argument('--shapenet-rendering-root', type=str, help='Root directory of the ShapeNet Rendering dataset.')
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/train.py]: Same than for eval.py

File: examples/ImageRecon/GEOmetrics/utils.py [hunk 6]
```diff
@@ -14,489 +14,485 @@
...
-		data_meta['y1s'].append(y1s.long().unsqueeze(0))
-		data_meta['y2s'].append(y2s.long().unsqueeze(0))
-
-	for key in data_meta:
-		data_meta[key] = torch.cat(data_meta[key], dim=0)
-	return data_meta
-
-
-
+def collate_fn(batch):
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/utils.py:31]: I wonder if this shouldn't be part of our core library (or having collate method associated to each dataset class)

File: examples/ImageRecon/GEOmetrics/utils.py [hunk 7]
```diff
@@ -14,489 +14,485 @@
...
 
 
-		features = (A*C*G + H*D*A + G*E*B + B*F*H).permute(1,0)
+        features = (A*C*G + H*D*A + G*E*B + B*F*H).permute(1, 0)
 
-		if full_features is None: full_features = features
-		else: full_features = torch.cat((full_features, features), dim = 1)
- 
-	return full_features
+        if full_features is None:
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/utils.py]: Not really important but `full_features = features if full_features if None else torch.cat((full_features, features), dim=1)` is more pythonic

**Review by Caenorst — COMMENTED** [review]
just few little comments, otherwise LGTM

File: examples/ImageRecon/GEOmetrics/architectures.py (hunk 2 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/architectures.py]: While you're correct, I'll leave this as is - defining side_2 makes it a clearer what is being concatenated.

File: examples/ImageRecon/GEOmetrics/eval.py (hunk 3 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/eval.py]: Happy to go with yours - it's shorter and easier to write.

File: examples/ImageRecon/GEOmetrics/eval.py (hunk 4 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/eval.py]: Done.

File: examples/ImageRecon/GEOmetrics/train.py (hunk 5 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/train.py]: Done.

File: examples/ImageRecon/GEOmetrics/utils.py (hunk 6 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/utils.py:31]: I agree. Would also address issue #95. Let's put that work in a subsequent PR.

File: examples/ImageRecon/GEOmetrics/utils.py (hunk 7 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/utils.py]: Done.

File: examples/ImageRecon/GEOmetrics/README.md (hunk 1 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/README.md:22]: Done.

File: examples/ImageRecon/GEOmetrics/utils.py [hunk 8]
```diff
@@ -14,489 +14,467 @@
...
+    loss += kal.metrics.mesh.edge_length(meshes['update'][2])
+    return loss
 
 
+def loss_lap(meshes): 
+    loss = .3 * kal.metrics.mesh.laplacian_loss(meshes['init'][0], meshes['update'][0])
+    loss += kal.metrics.mesh.laplacian_loss(meshes['init'][1], meshes['update'][1])
+    loss += kal.metrics.mesh.laplacian_loss(meshes['init'][2], meshes['update'][2])
+
+    loss += torch.sum((meshes['init'][1].vertices - meshes['update'][1].vertices)**2, 1).mean() * .0666
```
Comment (Caenorst) [inline examples/ImageRecon/GEOmetrics/utils.py]: spaces missing around "**"

File: kaolin/models/SimpleGCN.py [hunk 9]
```diff
@@ -55,31 +51,30 @@ def __init__(self, in_features, out_features, bias=True):
...
     def forward(self, input, adj):
-        
         support = torch.mm(input, self.weight1)
-        side_len = max(support.shape[1]//3, 2)
+        side_len = max(support.shape[1] // 3, 2)
         if adj.type() == 'torch.cuda.sparse.FloatTensor': 
-         
-            norm = torch.sparse.mm(adj,torch.ones((support.shape[0], 1)).cuda())
-            normalized_support = support[:, :side_len] /norm
+
```
Comment (Caenorst) [inline kaolin/models/SimpleGCN.py]: Unecessary line break

File: kaolin/models/GraphResNet.py [hunk 10]
```diff
@@ -31,8 +24,8 @@ class GraphResNet(nn.Module):
 
     """
 
-    def __init__(self, input_features, hidden = 192, output_features = 3):
-        super(G_Res_Net, self).__init__()
+    def __init__(self, input_features, hidden=192, output_features=3):
+        super(GraphResNet, self).__init__()
         self.gc1 = SimpleGCN(input_features, hidden)
         self.gc2 = SimpleGCN(hidden, hidden)
         self.gc3 = SimpleGCN(hidden , hidden)
```
Comment (Caenorst) [inline kaolin/models/GraphResNet.py]: Unecessary space before ","

File: kaolin/transforms/pointcloudfunc.py [hunk 11]
```diff
@@ -168,19 +168,14 @@ def realign(src: Union[torch.Tensor, PointCloud],
...
-    src = src - src.mean(-2).unsqueeze(-2)
-    src = ( (tgt_max - tgt_min) / (src_max - src_min + EPS) ) * src
-    # Undo the centering translation, and return the result.
-    return src + tgt.mean(-2).unsqueeze(-2)
+    src_min, _ = src.min(-2, keepdim=True)
+    src_max, _ = src.max(-2, keepdim=True)
+    tgt_min, _ = tgt.min(-2, keepdim=True)
+    tgt_max, _ = tgt.max(-2, keepdim=True)
+
+    src = ((tgt_max - tgt_min) / (src_max - src_min + EPS)) * src
```
Comment (Caenorst) [inline kaolin/transforms/pointcloudfunc.py]: Not sure about which one is faster, but I believe you can also do:
`src = ((src - src_min) / (src_max - src_min + EPS)) * (tgt_max - tgt_min) + tgt_min`

Instead of those two lines that would replace the `max` by a broadcast_add`. If you had the chance to test.

**Review by Caenorst — DISMISSED** [review]
Not easy to review because it's a lot of format changes as well as actual fixes but LGTM.

**Comment by Caenorst** [comment]
Seems like your signature is missing on the first commit.

File: examples/ImageRecon/GEOmetrics/utils.py (hunk 8 above)
Comment (Jean-Francois-Lafleche) [reply examples/ImageRecon/GEOmetrics/utils.py]: It's pretty common to not put spaces around the power operator (because it's the highest priority operation, and because it is closer to what you'd write on paper) from the projects I've come across, but I've checked out Pytorch and they seem add spaces, so I will go ahead an change. Note that Flake8 does not flag this as a format error, we might want to add this to the linter config to enforce it.

File: kaolin/models/SimpleGCN.py (hunk 9 above)
Comment (Jean-Francois-Lafleche) [reply kaolin/models/SimpleGCN.py]: Fixed.

File: kaolin/models/GraphResNet.py (hunk 10 above)
Comment (Jean-Francois-Lafleche) [reply kaolin/models/GraphResNet.py]: Missed those, thanks, good catch.

**Comment by Caenorst** [comment]
I don't see the latest fixed on the changes (due to squash + force push ?), my comments doesn't appear to be fixed ?

File: kaolin/transforms/pointcloudfunc.py (hunk 11 above)
Comment (Jean-Francois-Lafleche) [reply kaolin/transforms/pointcloudfunc.py]: Did some benchmarking: Perf between the two are very similar (your proposed version is slightly slower on CPU and slightly faster on GPU). Given that and the fact that I think your proposed version is easier to read, I've applied the change. Thank you!

**Review by Caenorst — APPROVED** [review]
Thanks for the fixes
